File: WebFt30/src/app/core/grpc/services/landmarks.service.ts
import { inject, Injectable } from '@angular/core';
import { GrpcWebFetchTransport } from '@protobuf-ts/grpcweb-transport';
import { AuthInterceptor } from '../auth.interceptor';
import { GrpcUtils } from '../grpc.utils';
import * as grpc from 'src/grpc-generated';
import { Observable } from 'rxjs';
import { ColoredLandmark } from '../../store/models/ft30/colored-landmark';
import { LandmarksMapping } from '../../store/mapping/landmarks-mapping';

@Injectable({
  providedIn: 'root'
})
export class LandmarksService {
  private authInterceptor = inject(AuthInterceptor);
  private client: grpc.LandmarksClient;

  constructor() {
    const interceptors = [this.authInterceptor.toRpcInterceptor()];
    const options = GrpcUtils.getGrpcOptions(GrpcUtils.getApiUrl(), interceptors);
    const transport = new GrpcWebFetchTransport(options);
    this.client = new grpc.LandmarksClient(transport);
  }

  getLandmarksModel(landmarksModelId: string): Observable<grpc.GetLandmarksModelResponse> {
    const request: grpc.GetLandmarksModelRequest = { landmarksModelId };
    return GrpcUtils.unaryToObservable(
      this.client.getLandmarksModel.bind(this.client),
      request,
      {}
    );
  }

  deleteLandmarksModel(landmarksModelId: string): Observable<grpc.DeleteLandmarksModelResponse> {
    const request: grpc.DeleteLandmarksModelRequest = { landmarksModelId };
    return GrpcUtils.unaryToObservable(
      this.client.deleteLandmarksModel.bind(this.client),
      request,
      {}
    );
  }

  createLandmarksModel(
    landmarksModelId: string,
    traceId: string
  ): Observable<grpc.CreateLandmarksModelResponse> {
    const request: grpc.CreateLandmarksModelRequest = {
      landmarksModelId,
      traceId
    };
    return GrpcUtils.unaryToObservable(
      this.client.createLandmarksModel.bind(this.client),
      request,
      {}
    );
  }

  updateLandmarksModel(
    landmarksModelId: string,
    changedLandmark: ColoredLandmark | undefined,
    isFilterOn: boolean | undefined
  ) {
    const request: grpc.UpdateLandmarksModelRequest = {
      landmarksModelId,
      changedLandmark: changedLandmark
        ? LandmarksMapping.toGrpcColoredLandmark(changedLandmark!)
        : undefined,
      isFilterOn
    };
    return GrpcUtils.unaryToObservable(
      this.client.updateLandmarksModel.bind(this.client),
      request,
      {}
    );
  }

  clearLandmarksModel(landmarksModelId: string) {
    const request: grpc.ClearLandmarksModelRequest = { landmarksModelId };
    return GrpcUtils.unaryToObservable(
      this.client.clearLandmarksModel.bind(this.client),
      request,
      {}
    );
  }
}

File: Backend/src/Fibertest30.Api/Services/LandmarksService.cs
using Grpc.Core;
using MediatR;

namespace Fibertest30.Api;

public class LandmarksService(ISender mediator) : Landmarks.LandmarksBase
{
    public override async Task<GetLandmarksModelResponse> GetLandmarksModel(GetLandmarksModelRequest request, ServerCallContext context)
    {
        var landmarksModel =
            await mediator.Send(new GetLandmarksModelQuery(Guid.Parse(request.LandmarksModelId)),
                context.CancellationToken);

        return new GetLandmarksModelResponse() { LandmarksModel = landmarksModel.ToProto() };
    }

    public override async Task<CreateLandmarksModelResponse> CreateLandmarksModel(CreateLandmarksModelRequest request,
        ServerCallContext context)
    {
        await mediator.Send(
            new CreateLandmarksModelCommand(Guid.Parse(request.LandmarksModelId),
                Guid.Parse(request.TraceId)), context.CancellationToken);

        return new CreateLandmarksModelResponse();
    }

    public override async Task<UpdateLandmarksModelResponse> UpdateLandmarksModel(UpdateLandmarksModelRequest request,
        ServerCallContext context)
    {
        await mediator.Send(
            new UpdateLandmarksModelCommand(Guid.Parse(request.LandmarksModelId), 
                request.ChangedLandmark?.FromProto(), request.HasIsFilterOn ? request.IsFilterOn : null));
        return new UpdateLandmarksModelResponse();
    }

    public override async Task<DeleteLandmarksModelResponse> DeleteLandmarksModel(DeleteLandmarksModelRequest request,
        ServerCallContext context)
    {
        await mediator.Send(new DeleteLandmarksModelCommand(Guid.Parse(request.LandmarksModelId)),
            context.CancellationToken);

        return new DeleteLandmarksModelResponse();
    }

    public override async Task<CancelOneLandmarkChangesResponse> CancelOneLandmarkChanges(CancelOneLandmarkChangesRequest request,
        ServerCallContext context)
    {
        await mediator.Send(new CancelOneLandmarkChangesCommand(
            Guid.Parse(request.LandmarksModelId), request.Row), context.CancellationToken);

        return new CancelOneLandmarkChangesResponse();
    }

    public override async Task<ClearLandmarksModelResponse> ClearLandmarksModel(ClearLandmarksModelRequest request,
        ServerCallContext context)
    {
        await mediator.Send(new ClearLandmarksModelCommand(Guid.Parse(request.LandmarksModelId)),
            context.CancellationToken);

        return new ClearLandmarksModelResponse();
    }

    public override async Task<ApplyLandmarkChangesResponse> ApplyLandmarkChanges(ApplyLandmarkChangesRequest request, 
        ServerCallContext context)
    {
        await mediator.Send(new ApplyLandmarkChangesCommand(Guid.Parse(request.LandmarksModelId)),
            context.CancellationToken);

        return new ApplyLandmarkChangesResponse();
    }
}